src/rust/src/history_ffi.rs

```diff
@@ -0,0 +1,264 @@
+use std::{convert::TryFrom, slice};
+
+use libc::{c_uchar, size_t};
+use zcash_history::{Entry as MMREntry, Tree as MMRTree, Version, V1, V2};
+use zcash_primitives::consensus::BranchId;
+
+/// Switch the tree version on the epoch it is for.
+fn dispatch<T>(cbranch: u32, v1: impl FnOnce() -> T, v2: impl FnOnce() -> T) -> T {
+    match BranchId::try_from(cbranch).unwrap() {
+        BranchId::Sprout
+        | BranchId::Overwinter
+        | BranchId::Sapling
+        | BranchId::Heartwood
+        | BranchId::Canopy => v1(),
+        _ => v2(),
+    }
+}
+
+fn construct_mmr_tree<V: Version>(
+    // Consensus branch id
+    cbranch: u32,
+    // Length of tree in array representation
+    t_len: u32,
+
+    // Indices of provided tree nodes, length of p_len+e_len
+    ni_ptr: *const u32,
+    // Provided tree nodes data, length of p_len+e_len
+    n_ptr: *const [c_uchar; zcash_history::MAX_ENTRY_SIZE],
+
+    // Peaks count
+    p_len: size_t,
+    // Extra nodes loaded (for deletion) count
+    e_len: size_t,
+) -> Result<MMRTree<V>, &'static str> {
+    let (indices, nodes) = unsafe {
+        (
+            slice::from_raw_parts(ni_ptr, p_len + e_len),
+            slice::from_raw_parts(n_ptr, p_len + e_len),
+        )
+    };
+
+    let mut peaks: Vec<_> = indices
+        .iter()
+        .zip(nodes.iter())
+        .map(
+            |(index, node)| match MMREntry::from_bytes(cbranch, &node[..]) {
+                Ok(entry) => Ok((*index, entry)),
+                Err(_) => Err("Invalid encoding"),
+            },
+        )
+        .collect::<Result<_, _>>()?;
+    let extra = peaks.split_off(p_len);
+
+    Ok(MMRTree::new(t_len, peaks, extra))
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_mmr_append(
+    // Consensus branch id
+    cbranch: u32,
+    // Length of tree in array representation
+    t_len: u32,
+    // Indices of provided tree nodes, length of p_len
+    ni_ptr: *const u32,
+    // Provided tree nodes data, length of p_len
+    n_ptr: *const [c_uchar; zcash_history::MAX_ENTRY_SIZE],
+    // Peaks count
+    p_len: size_t,
+    // New node pointer
+    nn_ptr: *const [u8; zcash_history::MAX_NODE_DATA_SIZE],
+    // Return of root commitment
+    rt_ret: *mut [u8; 32],
+    // Return buffer for appended leaves, should be pre-allocated of ceiling(log2(t_len)) length
```

Comment:
`log2(t_len)` is only defined when `t_len > 1`. Document this (or any other) restriction on `t_len`.